core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java

```diff
@@ -102,48 +100,146 @@
         this.copyResourceNotification = copyResourceNotification;
         this.classTransformers = classTransformers;
+        this.testSupport = testSupport;
+        if (testSupport != null) {
+            testSupport.addListener(new TestListener() {
+                @Override
+                public void testsEnabled() {
+                    if (!firstTestScanComplete) {
+                        checkForChangedTestClasses(true);
+                        firstTestScanComplete = true;
+                    }
+                    startTestScanningTimer();
+                }
+
+                @Override
+                public void testsDisabled() {
+                    synchronized (RuntimeUpdatesProcessor.this) {
+                        if (timer != null) {
+                            timer.cancel();
+                            timer = null;
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    public TestSupport getTestSupport() {
+        return testSupport;
     }
 
     @Override
     public Path getClassesDir() {
         //TODO: fix all these
         for (DevModeContext.ModuleInfo i : context.getAllModules()) {
-            return Paths.get(i.getClassesPath());
+            return Paths.get(i.getMain().getClassesPath());
         }
         return null;
     }
 
     @Override
     public List<Path> getSourcesDir() {
-        return context.getAllModules().stream().flatMap(m -> m.getSourcePaths().stream()).map(Paths::get).collect(toList());
+        return context.getAllModules().stream().flatMap(m -> m.getMain().getSourcePaths().stream()).map(Paths::get)
+                .collect(toList());
+    }
+
+    private Timer startTestScanningTimer() {
+        synchronized (this) {
+            if (timer == null) {
+                timer = new Timer("Test Compile Timer", true);
+                timer.schedule(new TimerTask() {
+                    @Override
+                    public void run() {
+                        periodicTestCompile();
+                    }
+                }, 1, 1000);
+            }
+        }
+        return timer;
+    }
+
+    private void periodicTestCompile() {
+        //noop if already scanning
+        if (scanLock.tryLock()) {
+            try {
+                ClassScanResult changedTestClassResult = compileTestClasses();
+                ClassScanResult changedApp = checkForChangedClasses(compiler, DevModeContext.ModuleInfo::getMain, false, test);
+                Set<String> filesChanged = checkForFileChange(DevModeContext.ModuleInfo::getMain, test);
+                boolean configFileRestartNeeded = filesChanged.stream().map(watchedFilePaths::get)
+                        .anyMatch(Boolean.TRUE::equals);
+                ClassScanResult merged = ClassScanResult.merge(changedTestClassResult, changedApp);
+                if (configFileRestartNeeded) {
+                    if (compileProblem != null) {
```

Comment:
This could be fine but just to clarify it is. We would never have a `compileProblem` we wanted to report unless either the `configFileRestartNeeded` or the `ClassScanResult.isChanged()`?